activemq-unit-tests/src/test/java/org/apache/activemq/jms2/ActiveMQJMS2MessageTypesTest.java

```diff
@@ -0,0 +1,582 @@
+            assertNotNull(destination);
+            jmsContext.start();
+
+            Message message = ActiveMQJMS2TestSupport.generateMessage(jmsContext, messageType, messagePayload);
+            boolean caught = false;
+            try {
+                ActiveMQJMS2TestSupport.sendMessage(jmsContext, destination, message, null, 99, null, null, null, null, null, null, null);
+                fail("IlegalStateRuntimeException expected");
+            } catch (IllegalStateRuntimeException e) {
+                assertEquals("unknown delivery mode: 99", e.getMessage());
+                caught = true;
+            }
+            assertTrue(caught);
+        } catch (Exception e) {
+            fail(e.getMessage());
+        }
+    }
+
+    @Test
+    public void testMessagePriority() {
+
+        try(JMSContext jmsContext = activemqConnectionFactory.createContext("admin", "admin", Session.AUTO_ACKNOWLEDGE)) {
+            assertNotNull(jmsContext);
+            Destination destination = ActiveMQJMS2TestSupport.generateDestination(jmsContext, destinationType, methodNameDestinationName);
+            assertNotNull(destination);
+            JMSConsumer jmsConsumer = jmsContext.createConsumer(destination);
+            assertNotNull(jmsConsumer);
+            jmsContext.start();
+
+            Message message = ActiveMQJMS2TestSupport.generateMessage(jmsContext, messageType, messagePayload);
+
+            List<String> sentMessageIds = new LinkedList<>();
+            for(int priority=0; priority<10; priority++) {
+                sentMessageIds.add(ActiveMQJMS2TestSupport.sendMessage(jmsContext, destination, message, null, null, null, null, null, null, null, priority, null));
+            }
+
+            Message recvMessage = null;
+            List<Message> recvMessages = new LinkedList<>();
+            int loopCount = 0;
+            int maxLoops = 50;
+            boolean done = false;
+            do {
+                recvMessage = jmsConsumer.receive(500l);
+                if(recvMessage != null) {
+                    recvMessages.add(recvMessage);
+                }
+
+                if(recvMessages.size() == 10) {
+                    done = true;
+                }
+                loopCount++;
+            } while (loopCount <= maxLoops && !done);
+
+            int foundCount = 0;
+            for(int validatedPriority=0;validatedPriority<10;validatedPriority++) {
+                for(javax.jms.Message tmpMessage : recvMessages) {
+                    if(tmpMessage.getJMSPriority() == validatedPriority) {
+                        ActiveMQJMS2TestSupport.validateMessagePriority(tmpMessage, messageType, messagePayload, validatedPriority);
+                        foundCount++;
+                    }
+                }
+            }
+            assertEquals(Integer.valueOf(10), Integer.valueOf(foundCount));
```

Comment:
Adding the message priorities to a TreeSet or sopmething and asserting it contains 0 to 9 might be simpler and clearer. Or you could create the consumer after they were sent, and jsut expect them in reverse (i.e priority) order and validate on arrival.